File: chess-frontend/src/utils/openingDetection.js
// Shared opening book and detection utility
// Extracted from videoExportUtils.js for reuse across GameReview, video export, etc.

export const OPENING_BOOK = [
  { moves: ['e4','e5','Nf3','Nc6','Bb5'], name: 'Ruy Lopez' },
  { moves: ['e4','e5','Nf3','Nc6','Bc4'], name: 'Italian Game' },
  { moves: ['e4','e5','Nf3','Nc6','d4'], name: 'Scotch Game' },
  { moves: ['e4','c5','Nf3','d6','d4'], name: 'Sicilian Najdorf' },
  { moves: ['e4','c5','Nf3','Nc6','d4'], name: 'Sicilian Open' },
  { moves: ['d4','d5','c4','e6','Nc3'], name: "Queen's Gambit Declined" },
  { moves: ['d4','d5','c4','dxc4'], name: "Queen's Gambit Accepted" },
  { moves: ['d4','Nf6','c4','g6','Nc3'], name: "King's Indian Defense" },
  { moves: ['d4','Nf6','c4','e6','Nc3','Bb4'], name: 'Nimzo-Indian Defense' },
  { moves: ['e4','e5','Nf3','Nf6'], name: "Petrov's Defense" },
  { moves: ['e4','e5','f4'], name: "King's Gambit" },
  { moves: ['e4','e5','Nc3'], name: 'Vienna Game' },
  { moves: ['d4','Nf6','c4','c5'], name: 'Benoni Defense' },
  { moves: ['e4','c5'], name: 'Sicilian Defense' },
  { moves: ['e4','e6'], name: 'French Defense' },
  { moves: ['e4','c6'], name: 'Caro-Kann Defense' },
  { moves: ['e4','d5'], name: 'Scandinavian Defense' },
  { moves: ['e4','g6'], name: 'Modern Defense' },
  { moves: ['e4','d6'], name: 'Pirc Defense' },
  { moves: ['e4','Nf6'], name: "Alekhine's Defense" },
  { moves: ['d4','d5','c4'], name: "Queen's Gambit" },
  { moves: ['d4','f5'], name: 'Dutch Defense' },
  { moves: ['d4','Nf6','c4','g6'], name: "King's Indian" },
  { moves: ['Nf3','d5','g3'], name: "King's Indian Attack" },
  { moves: ['c4'], name: 'English Opening' },
  { moves: ['Nf3'], name: 'Reti Opening' },
  { moves: ['e4','e5'], name: "King's Pawn Game" },
  { moves: ['d4','d5'], name: "Queen's Pawn Game" },
  { moves: ['d4','Nf6'], name: 'Indian Defense' },
].sort((a, b) => b.moves.length - a.moves.length);

export const detectOpening = (sanMoves) => {
  for (const opening of OPENING_BOOK) {
    if (opening.moves.length <= sanMoves.length &&
        opening.moves.every((m, i) => sanMoves[i] === m)) {
      return opening.name;
    }
  }
  return null;
};

File: chess99-android/app/src/main/java/com/chess99/engine/OpeningDetection.kt
package com.chess99.engine

/**
 * Detects chess opening from a sequence of SAN moves.
 * Ported from chess-frontend/src/utils/openingDetection.js.
 *
 * Matches the longest opening from the book that aligns with the game's moves.
 * Sorted by length descending so the most specific match wins.
 */

private data class Opening(val moves: List<String>, val name: String)

private val OPENING_BOOK: List<Opening> = listOf(
    Opening(listOf("e4","e5","Nf3","Nc6","Bb5"), "Ruy Lopez"),
    Opening(listOf("e4","e5","Nf3","Nc6","Bc4"), "Italian Game"),
    Opening(listOf("e4","e5","Nf3","Nc6","d4"), "Scotch Game"),
    Opening(listOf("e4","c5","Nf3","d6","d4"), "Sicilian Najdorf"),
    Opening(listOf("e4","c5","Nf3","Nc6","d4"), "Sicilian Open"),
    Opening(listOf("d4","d5","c4","e6","Nc3"), "Queen's Gambit Declined"),
    Opening(listOf("d4","d5","c4","dxc4"), "Queen's Gambit Accepted"),
    Opening(listOf("d4","Nf6","c4","g6","Nc3"), "King's Indian Defense"),
    Opening(listOf("d4","Nf6","c4","e6","Nc3","Bb4"), "Nimzo-Indian Defense"),
    Opening(listOf("e4","e5","Nf3","Nf6"), "Petrov's Defense"),
    Opening(listOf("e4","e5","f4"), "King's Gambit"),
    Opening(listOf("e4","e5","Nc3"), "Vienna Game"),
    Opening(listOf("d4","Nf6","c4","c5"), "Benoni Defense"),
    Opening(listOf("e4","c5"), "Sicilian Defense"),
    Opening(listOf("e4","e6"), "French Defense"),
    Opening(listOf("e4","c6"), "Caro-Kann Defense"),
    Opening(listOf("e4","d5"), "Scandinavian Defense"),
    Opening(listOf("e4","g6"), "Modern Defense"),
    Opening(listOf("e4","d6"), "Pirc Defense"),
    Opening(listOf("e4","Nf6"), "Alekhine's Defense"),
    Opening(listOf("d4","d5","c4"), "Queen's Gambit"),
    Opening(listOf("d4","f5"), "Dutch Defense"),
    Opening(listOf("d4","Nf6","c4","g6"), "King's Indian"),
    Opening(listOf("Nf3","d5","g3"), "King's Indian Attack"),
    Opening(listOf("c4"), "English Opening"),
    Opening(listOf("Nf3"), "Reti Opening"),
    Opening(listOf("e4","e5"), "King's Pawn Game"),
    Opening(listOf("d4","d5"), "Queen's Pawn Game"),
    Opening(listOf("d4","Nf6"), "Indian Defense"),
).sortedByDescending { it.moves.size }

fun detectOpening(sanMoves: List<String>): String? {
    for (opening in OPENING_BOOK) {
        if (opening.moves.size <= sanMoves.size &&
            opening.moves.indices.all { i -> sanMoves[i] == opening.moves[i] }
        ) {
            return opening.name
        }
    }
    return null
}
